examples/voting-locker/locker.rho

```diff
@@ -0,0 +1,148 @@
+
+TODO: work out how seed, arg are packaged for signature. Currently,
+the signature checking algorigthm is a toy: sig == pk ++ pk.
+
+The demo below uses the Locker for two different roles in a voting
+protocol: the chair and a voter. The voter's seed process creates an
+inbox channel and shares a write-only bundle of it with the chair via
+the registry.  The chair sends a right to vote on this channel and the
+voter then excercises the vote.
+*TODO: more than one voter*
+
+The demo is also factored for use by off-chain components that pass
+plain data in and out (with the exception of return channels).
+*TODO: insertArbitrary(*chair, ...)*
+
+*/
+
+new export(`export:`), Locker, verify, toyVerify,
+trace(`rho:io:stderr`), stdout(`rho:io:stdout`),
+insertArbitrary(`rho:registry:insertArbitrary`), lookup(`rho:registry:lookup`)
+in {
+  export!(bundle+{*Locker}) |
+
+  // WARNING: if sig is the wrong size, we get a java.lang.RuntimeException,
+  // and our nonceCh is stuck.
+  contract verify(
+    @{sig /\ ByteArray},  // ed25519 signature of blakeb256Hash of ... TODO
+    @{pk /\ ByteArray},   // public key
+    @{nonce /\ Int},      // claimed next nonce
+    nonceCh,              // real nonce storage
+    @data,                // data to be signed (along with nonce)
+    eject,                // failure path (bad nonce or signature)
+    ack                   // sig OK path
+  ) = {
+    // trace!({"verify sig": sig, "pk": pk, "nonce": nonce}) |
+    for(@prevNonce <- nonceCh) {
+      // trace!({"prevNonce": prevNonce}) |
+      if (nonce == (prevNonce + 1)) {
+        new result, hashOut in {
+          // ISSUE: how to package nonce, data? tuples are nice,
+          // but they don't map easily to JSON.
+          @"blake2b256Hash"!((nonce, data).toByteArray(), *hashOut) |
+          for(@hash <- hashOut) {
+            // trace!({"hash": hash}) |
+            @"ed25519Verify"!(hash, sig, pk, *result) | for(@r <- result) {
+              // trace!({"verify": r}) |
+              if (r) { nonceCh!(nonce) | ack!(Nil) }
+              else { nonceCh!(prevNonce) | eject!("bad ed25519 signature") }
+            }
+          }
+        }
+      } else {
+        // ISSUE: having the locker's nickname in the diagnostic would be nice
+        nonceCh!(prevNonce) | eject!({"expected nonce": prevNonce + 1})
+      }
+    }
+  } |
+
+  // toy signature verification: checks that sig == pk ++ pk
+  contract toyVerify(
+    @{sig /\ ByteArray}, @{pk /\ ByteArray}, @{nonce /\ Int}, nonceCh,
+    @data,
+    eject, ack) = {
+    // trace!({"verify sig": sig, "pk": pk, "data": data}) |
+    for(@prevNonce <- nonceCh) {
+      // trace!({"prevNonce": prevNonce}) |
+      if (nonce == (prevNonce + 1)) {
+        if (pk ++ pk == sig) { nonceCh!(nonce) | ack!(Nil) }
+        else { nonceCh!(prevNonce) | eject!("bad ++ sig") }
+      } else {
+        nonceCh!(prevNonce) | eject!({"expected nonce": prevNonce + 1})
+      }
+    }
+  } |
+
+  contract Locker(
```

Comment:
This nonce-based implementation will require a pretty complicated nonce management, considering the parallelism of the rholang programs.

I have the following alternative proposal for the Locker contract (probably not syntactic correct), which simplified the code and eliminates the need for any nonce-management:

```
contract Locker(
    @{label /\ String},
    @{pubKey /\ ByteArray},
    @seedCode, seedEntry,
    return) = {
    new self, seedCh in {
      return!(*self) |
      
      seedCode |
      seedEntry!(*seedCh) | for (seed <- seedCh) {
        contract self(@"getSeedProxy", @{sig /\ ByteArray}, return) = { 
          verify!(sig, pubKey, ("getSeedProxy", *return), *eject, *okCh) |
          for(_ <- okCh) {
            new proxy in {
              for (msg <- proxy) { seed ! (msg)} |
              return ! (proxy)
            }
          }
        }
      }
    }
  }```

And now the Locker would be used like this:

```
Locker!("getSeedProxy", signature, return) | 
for (proxy <- return) { 
  // use the proxy
}```